---
title: "Visualizing Bile Acids Proteomics results"
output:
  pdf_document: 
    highlight: tango
  html_notebook: default
---

```{r lib-setup, echo=TRUE, warning=FALSE, message=FALSE}
library(clusterProfiler)
library(org.Hs.eg.db)
library(tidyverse)
library(readxl)
library(ggprism)
library(patchwork)
library(ggtext)
library(glue)
library(umap)
```

Set up parameters - like colors 
```{r setup}
div_pal <- rcartocolor::carto_pal(7, "Geyser")
```


```{r import}
mq <- read_tsv(here::here("normalyzer_limma_de_results_annotated.tsv"))
meta <- read_tsv(here::here("norm_meta.tsv"))
mq_res <- mq %>% 
  select(contains("Log"), contains("Adj"), contains("PValue"), contains("Majority"), contains("Gene"))
mq_mat <- mq %>% 
  select(`Gene names`, `Majority protein IDs`, matches("(C|L)[1-6](c|m)"))
#Convert the results table into a list and unify the columns names. 
#Invert the foldChanges, so that comparisons are L-C instead of C-L.
#Reduce Gene names to just one gene for analysis
mq_res_list <- list(select(mq_res, -contains("Cc-Lc")),
                    select(mq_res, -contains("Cm-Lm"))) %>% 
  set_names(nm = c("Membrane", "Cyto")) %>% 
  map(rename_with, ~str_remove(.,"C(c|m)-L(c|m)_") ,matches("C(c|m)-L(c|m)_")) %>% 
  map(mutate, across(contains("log2"), ~ -.x)) %>% 
  map(separate, `Gene names`, into = letters [1:2], extra = "drop", sep = ";") %>% 
  map(mutate, Genes = case_when(
    str_detect(a, "orf") & !is.na(b) & str_detect(b, "orf", negate = T) ~ b,
    TRUE ~ a
  ), Gene_rows = Genes) 
```
# Venn diagram

```{r venn}
#Filter results to only include DA proteins
da_list <- mq_res_list %>% 
  map(filter, AdjPVal < 0.05, abs(log2FoldChange) > 1)
#Pull just the Protein IDs for DA proteins
da_genes <- da_list %>% map(pull, `Majority protein IDs`) %>% 
  set_names(nm = c("Membrane-bound Proteins", "Cytosolic Proteins"))
ggvenn::ggvenn(da_genes, text_size = 10, set_name_size = 6,
               fill_color = c("indianred", "royalblue1"), show_percentage = F)
#ggsave(here::here("Euler diagram DA proteins.png"), units = "in", dpi = 720, width = 8, height = 6)
```
Get identities of common DE proteins
```{r intersect-de}
common_top <- da_genes %>% 
  reduce(intersect)
mq_intersect_top <- mq_res %>% 
  filter(`Majority protein IDs` %in% common_top) 
#write_csv(mq_intersect_top, here::here("Shared DE proteins.csv"))
```


# Volcano plots
```{r volcano, fig.width=5, fig.height=3.3}
#Add categories for signifcance
mq_res_volc_list <- mq_res_list %>% 
  map(mutate, 
      Significance = case_when(
        AdjPVal >= 0.05 | is.na(AdjPVal) ~ "Not Significant",
        AdjPVal < 0.05 & log2FoldChange > 0 ~ "Up (q<0.05)",
        AdjPVal < 0.05 & log2FoldChange < 0 ~ "Down (q<0.05)",
      )) %>% 
  map(mutate, 
      Significance = fct_relevel(Significance,"Not Significant","Up (q<0.05)","Down (q<0.05)"))

volcano_list <- mq_res_volc_list %>% 
  map2(c("Membrane-bound Proteins", "Cytosolic Proteins"), 
       ~ ggplot(.x, aes(x = log2FoldChange, y = -log10(PValue))) +
         geom_point(aes(fill = Significance, color = Significance), shape = 21, alpha = 0.5, size = 1.5) +
         scale_color_manual(values = c("grey85",div_pal [c(1,7)])) +
         scale_fill_manual(values = c("grey85",div_pal [c(1,7)])) + 
         labs(x = bquote(~Log[2] ~ "fold change"), y = bquote(~-Log[10] ~ italic(P))) +
         theme_linedraw(base_size = 15) +
         theme(panel.grid = element_line(color =  "grey90"),
               panel.grid.major = element_blank(),
               panel.grid.minor = element_blank(),
               panel.border = element_rect(fill = NA, color = "black", linewidth = 0.9)))
walk(volcano_list, print)
#walk2(volcano_list, names(volcano_list), 
    # ~ggsave(here::here(str_c("final results/volcano_plot_",.y,".pdf")), plot = .x, width = 8, height = 6.66, dpi = 1080))
walk2(volcano_list, names(volcano_list), 
     ~ggsave(here::here(str_c("final results/fig. 3b small_volcano_plot_",.y,".pdf")), plot = .x, width = 5, height = 3.33, dpi = 2160))
#walk2(volcano_list, names(volcano_list), 
 #    ~ggsave(here::here(str_c("final results/fig. 3b small_volcano_plot_",.y,".svg")), plot = .x, width = 5, height = 3.33, dpi = 2160))
```


# Functional enrichment
```{r cluster-enrich}
set.seed(711)
#Get the whole human genome as universe
human_genom <- org.Hs.egSYMBOL
mapped_genes <- mappedRkeys(human_genom)
#Try to get symbols from mq_res as universe
universe_df <- mq_res %>% 
  select(Gene = `Gene names`) %>% 
  deframe() %>%  
  map(str_split, ";") %>% 
  flatten() %>% 
  flatten_chr()
h <- read.gmt(list.files(path = here::here("."),pattern = "h.all"))
c2_cp <- read.gmt(list.files(path = here::here("."),pattern = "c2.cp"))  
c2_kegg <- c2_cp %>% 
  filter(str_detect(term, "KEGG")) 
c2_react <- c2_cp %>% 
  filter(str_detect(term,"REACTOME"))
#Set up lists of genes for analysis
prots_list <- vector(mode = "list", length = 4) %>% 
  set_names(map_chr(cross2(c("Membrane-bound","Cytoplasmic"),c("Up","Down")), reduce, str_c, sep = "_"))
#Select fold change cut-off and run the loop
cutoff <- 1
for(i in seq_along(prots_list)) {
  if (str_detect(names(prots_list) [i], "Membrane")) {
    res_dummy <- da_list [["Membrane"]]
    if (str_detect(names(prots_list) [i], "Up")) {
      prots_list [[i]] <- pull(filter(res_dummy, AdjPVal < 0.05, log2FoldChange > cutoff), var = "Genes")
    } else {
      prots_list [[i]] <- pull(filter(res_dummy, AdjPVal < 0.05, log2FoldChange < -cutoff), var = "Genes")
    }
  } else {
    res_dummy <- da_list [["Cyto"]]
    if (str_detect(names(prots_list) [i], "Up")) {
      prots_list [[i]] <- pull(filter(res_dummy, AdjPVal < 0.05, log2FoldChange > cutoff), var = "Genes")
    } else {
      prots_list [[i]] <- pull(filter(res_dummy, AdjPVal < 0.05, log2FoldChange < -cutoff), var = "Genes")
    }
  }
}


enrich_h_list <- map(prots_list, ~enricher(.,
                     TERM2GENE = h,
                     universe = universe_df,
                     pAdjustMethod = "BH",
                     pvalueCutoff = 0.05,
                     qvalueCutoff = 0.05))
ego_bp <- map(prots_list, ~enrichGO(gene = .,
                OrgDb         = org.Hs.eg.db,
                universe = universe_df,
                keyType       = 'SYMBOL',
                ont           = "BP",
                pAdjustMethod = "BH",
                pvalueCutoff  = 0.05,
                qvalueCutoff  = 0.05))
ego_bp_suc <- map(keep(ego_bp, ~nrow(.) > 0),enrichplot::pairwise_termsim)
ego_bp2 <- ego_bp_suc %>% 
  map(~clusterProfiler::simplify(., cutoff=0.7, by="p.adjust", select_fun=min)) %>% 
  set_names(str_c(names(.), "GOBP", sep = "__"))

#Create a list for different C2 databases
prots_c2_list <- cross2(prots_list, list("KEGG" = c2_kegg, "REACTOME" = c2_react)) %>% 
  set_names(map_chr(cross2(names(prots_list),c("KEGG","REACTOME")), reduce, str_c, sep = "__"))

enrich_c2_list <- map(prots_c2_list, ~enricher(.x [[1]],
                     TERM2GENE = .x [[2]],
                     universe = universe_df,
                     pAdjustMethod = "BH",
                     pvalueCutoff = 0.05,
                     qvalueCutoff = 0.05))

```

Plot the functional enrichment results. Combine the up and down regulated pathways from the same database.
Calculate OddsRatio as [described here](https://support.bioconductor.org/p/89304/)

```{r cluster-plot-prep, fig.height=9, fig.width=7}
#Keep only groups with significant results
enrich_h_suc_list <- keep(enrich_h_list, ~nrow(.) > 0) %>% 
  set_names(str_c(names(.),"Hallmark",  sep = "__"))

#Process the dataframes 
enrich_df_list <- prepend(ego_bp2,enrich_h_suc_list) %>%   
  prepend(keep(enrich_c2_list, ~nrow(.) > 0)) %>%  
  map(as.data.frame) %>% 
  map(mutate, 
      num_gene_ratio = sapply(GeneRatio, function(x) eval(parse(text=x))),
      num_bg_ratio = sapply(BgRatio, function(x) eval(parse(text=x)))) %>% 
  map(mutate, OddsRatio = num_gene_ratio/num_bg_ratio)
enrich_df_list <- enrich_df_list %>% 
  map2(names(enrich_df_list), 
       ~mutate(.x, Direction = if_else(str_detect(.y,"Up"), "Up-regulated","Down-regulated")))
#Combine the dataframes for up and down regulation
enrich_df_comb_list <- vector(mode = "list", length = 1)
names(enrich_df_comb_list) <- "placeholder"
j <- 1
#Create a loop to combine all the dataframes
for (i in seq_along(enrich_df_list)) {
  #Extract the db and fraction name from the list name
  df_name <- names(enrich_df_list) [i]
  fract_name <- str_remove(str_extract(df_name, "[:alpha:]+[-_]"),"[-_]")
  db_name <- str_remove(str_extract(df_name, "__.+"),"__") 
  index <- intersect(str_which(names(enrich_df_list), fract_name),str_which(names(enrich_df_list), db_name))
  enrich_df_comb_list [[j]] <- reduce(enrich_df_list [index], bind_rows) 
  names(enrich_df_comb_list) [j] <- str_c(fract_name, db_name, sep = "__")
  j <- j+1
}  
#Remove duplicate entries and clean up names
enrich_df_comb_unq_list <- keep(enrich_df_comb_list, !base::duplicated(enrich_df_comb_list)) %>% 
  map(mutate, 
      clean_ID = str_remove(Description, "(KEGG)|(REACTOME)|(HALLMARK)"),
      clean_ID = str_replace_all(clean_ID, "_", " "),
      clean_ID = str_to_sentence(clean_ID),
      clean_ID = str_replace_all(clean_ID, "[Rr]na", "RNA"),
      clean_ID = str_replace_all(clean_ID, "RRNA", "rRNA"),
      clean_ID = str_replace_all(clean_ID, "E2f", "E2F"),
      clean_ID = str_replace_all(clean_ID, "[Aa]tp", "ATP"),
      clean_ID = str_replace_all(clean_ID, " i ", " I "),
      clean_ID = str_replace_all(clean_ID, " ii", " II"),
      clean_ID = str_replace_all(clean_ID, " IIi", " III"),
      clean_ID = str_replace_all(clean_ID, "[Jj]ak", "JAK"),
      clean_ID = str_replace_all(clean_ID, " [Ss]tat", " STAT"),
      clean_ID = str_replace_all(clean_ID, "[Tt]ca", "TCA"),
      clean_ID = str_replace_all(clean_ID, "[Aa]pc c", "APC C"),
      clean_ID = str_replace_all(clean_ID, "cdh", "CDH"),
      clean_ID = str_replace_all(clean_ID, "cdc", "CDC"),
      clean_ID = str_replace_all(clean_ID, "3s", "3S"),
      clean_ID = str_replace_all(clean_ID, "k11", "K11"),
      clean_ID = str_wrap(clean_ID, width = 30))
#Print the results into csv
#walk2(enrich_df_comb_unq_list, names(enrich_df_comb_unq_list),
 #     ~write_csv(.x, file = here::here(str_c("final results/enrichment_results/",.y,".csv"))))
```

Make the dotplots of the results
```{r cluster-plot, fig.height=4, fig.width=7}
#Finally can plot the results
#Plot some plots separately
same_dim_dots <- enrich_df_comb_unq_list %>% 
  map(mutate, 
      clean_ID = fct_reorder(clean_ID, OddsRatio),
      Mito = "No")
#Prep keywords to identify mitochondrial categories
mito_key <- c("Mitochondria", "NADH", "Energy", "Respiratory", "ATP", "Oxidative", "Complex",  "Respir", "ATP", "Oxidoreductase",
              "TCA")
#Loop through the keywords to identify the categories of relevance 
same_dim_dots <- same_dim_dots %>% 
  map(mutate, 
      Mito = map_lgl(clean_ID, ~any(str_detect(str_to_lower(.x),str_to_lower(mito_key)))))
#Create labels
same_dim_dots_labs_list <- same_dim_dots %>% 
  map(~glue_data(.x, "<span style='color: {if_else(Mito == 'TRUE', 'red', 'black')}'>{str_replace_all(clean_ID, '\\\\n','<br>')}</span>")) %>% 
  map2(same_dim_dots,~set_names(.x, nm = pull(.y, var = "clean_ID")))
#Select only plots with low category number to plot
small_index <- c(5,7)
low_cat_n_p <-  pmap(list(same_dim_dots [small_index], same_dim_dots_labs_list [small_index], names(same_dim_dots) [small_index]), 
                     ~ggplot(..1, aes(x = OddsRatio, y = clean_ID )) +
                       geom_point(aes(color = Direction, size = qvalue)) +
                       labs(y = element_blank(), title =str_remove(..3, ".+__"), size = "Significance") +
                       scale_color_manual(values = div_pal [c(7,1)]) + 
                       scale_size(range = c(8,4), labels = function(x) {str_c("q = ", x)}) +
                       scale_y_discrete(labels = ..2) +
                       guides(color = guide_legend(override.aes = list(size = 3))) +
                       coord_cartesian(clip = "off") +
                       theme_bw(base_size = 15) +
                       theme(panel.grid = element_blank(),
                             axis.text.y = element_markdown(lineheight = 1.1)))

#walk2(low_cat_n_p, names(low_cat_n_p), 
 #     ~ggsave(filename = here::here(str_c("final results/enrichment results ",.y,".png")), plot = .x, dpi = 1080, width = 8, height = 5))

low_cat_n_p [2]
#walk2(low_cat_n_p [2], names(low_cat_n_p) [2], 
 #     ~ggsave(filename = here::here(str_c("final results/fig. 3C enrichment results ",.y,".pdf")), plot = .x, dpi = 1080, width = 7, height = 4))
#walk2(low_cat_n_p [2], names(low_cat_n_p) [2], 
 #     ~ggsave(filename = here::here(str_c("final results/fig. 3C enrichment results ",.y,".svg")), plot = .x, dpi = 1080, width = 7, height = 4))
```
Now plot graphs of intermediate category number
```{r cluster-plot-intermedi, fig.height=6.66, fig.width=8}
inter_index <- c(-3,-4)
inter_cat_n_p <-pmap(list(same_dim_dots [inter_index], same_dim_dots_labs_list [inter_index], names(same_dim_dots) [inter_index]), 
                     ~ggplot(..1, aes(x = OddsRatio, y = clean_ID )) +
                       geom_point(aes(color = Direction, size = qvalue)) +
                       labs(y = element_blank(), title =..3, size = "Significance") +
                       scale_color_manual(values = div_pal [c(7,1)]) + 
                       scale_size(range = c(10,3), labels = function(x) {str_c("q = ", x)}) +
                       scale_y_discrete(labels = ..2) +
                       guides(color = guide_legend(override.aes = list(size = 3))) +
                       coord_cartesian(clip = "off") +
                       theme_bw(base_size = 14) +
                       theme(panel.grid = element_blank(),
                             axis.text.y = element_markdown(lineheight = 0.7)))
#walk2(inter_cat_n_p, names(inter_cat_n_p), 
 #     ~ggsave(filename = here::here(str_c("final results/enrichment results ",.y,".png")), plot = .x, dpi = 1080, width = 8, height = 6.66))
```

Finally plot graphs of large category number

```{r cluster-plot-large, fig.height=9.5, fig.width=10}
large_index <- c(3,4)
same_dim_dots <- same_dim_dots %>% 
  map(arrange, qvalue) %>% 
  map_at(large_index, slice_head, n=20) 

large_cat_n_p <- pmap(list(same_dim_dots [large_index], same_dim_dots_labs_list [large_index], names(same_dim_dots) [large_index]), 
                      ~ggplot(..1, aes(x = OddsRatio, y = clean_ID )) +
                        geom_point(aes(color = Direction, size = qvalue)) +
                        labs(y = element_blank(), title =..3, size = "Significance") +
                        scale_color_manual(values = div_pal [c(7,1)]) + 
                        scale_size(range = c(10,3), labels = function(x) {str_c("q = ", x)}) +
                        scale_y_discrete(labels = ..2) +
                        guides(color = guide_legend(override.aes = list(size = 3))) +
                        coord_cartesian(clip = "off") +
                        theme_bw(base_size = 14) +
                        theme(panel.grid = element_blank(),
                              axis.text.y = element_markdown(lineheight = 1.1)))
#walk2(large_cat_n_p, names(large_cat_n_p), 
 #    ~ggsave(filename = here::here(str_c("final results/enrichment results ",.y,".png")), plot = .x, dpi = 1080, width = 10, height = 9.5))
```


```{r cluster-plot-fig-3b, fig.height=5, fig.width=7}
#Change the membrane reactome figure for the manuscript according to requests
manu_df <- same_dim_dots [large_index [1]] %>%   
  rep(., 2) %>% 
  map2(c(5,7), ~slice_head(.x, n=.y))

manu_figs <- pmap(list(manu_df, same_dim_dots_labs_list [rep(large_index [1],2)], names(manu_df)), 
                      ~ggplot(..1, aes(x = OddsRatio, y = clean_ID )) +
                        geom_point(aes(color = Direction, size = qvalue)) +
                        labs(y = element_blank(), title =str_remove(..3, ".+__"), size = "Significance") +
                        scale_color_manual(values = div_pal [c(7,1)]) + 
                        scale_size(range = c(8,4), labels = function(x) {str_c("q = ", x)},
                                   breaks = scales::breaks_extended(n=3)) +
                        scale_y_discrete(labels = ..2) +
                        guides(color = guide_legend(override.aes = list(size = 3))) +
                        coord_cartesian(clip = "off") +
                        theme_bw(base_size = 15) +
                        theme(panel.grid = element_blank(),
                              axis.text.y = element_markdown(lineheight = 1.1)))
#pwalk(list(manu_figs, names(manu_figs),c(5,7)), 
 #    ~ggsave(filename = here::here(str_c("final results/fig 3D enrichment results ",..2," top ", ..3, ".pdf")), plot = ..1, dpi = 1080, width = 7, height = 5))
#pwalk(list(manu_figs, names(manu_figs),c(5,7)), 
 #    ~ggsave(filename = here::here(str_c("final results/fig 3D enrichment results ",..2," top ", ..3, ".svg")), plot = ..1, width = 7, height = 5))

```
Last step is to create a UMAP graph of the data
```{r umap, fig.width=6, fig.height=4}
#Import vsn-normalized counts
vsn_mq <- read_tsv(here::here(here::here("normalyzer/test_run/VSN-normalized.txt")))
vsn_mq_mat <- vsn_mq %>% 
  column_to_rownames(var = "Majority protein IDs") %>% 
  #vsn is akin to log-transform so replace missing values with 1s (log of 0 is 1)
  mutate(across(.fns = ~as.numeric(str_replace_na(.x, replacement = "1")))) %>% 
  as.matrix() %>% 
  t()
#Create list of fraction-based samples
vsn_mq_list <- map(set_names(c("c","m"), nm = c("Cyto","Membrane")), 
                   ~ vsn_mq_mat [str_detect(rownames(vsn_mq_mat), .x),])
#Run umap
#Reduce number of neighbours
custom_sets <- umap.defaults
custom_sets$n_neighbors <- 4
umap_fit_list <- map(vsn_mq_list, umap, config = custom_sets)
#Extract coordinates
umap_df_list <- umap_fit_list %>% 
  map(~.x$layout) %>% 
  map(as.data.frame) %>% 
  map(rename, UMAP1 = V1, UMAP2 = V2) %>% 
  map(rownames_to_column, var = "sample") %>% 
  map(mutate, Condition = if_else(str_detect(sample, "L"), "LCA", "Control"))

#Plot the results
umap_with_elements <- map2(umap_df_list, names(umap_df_list), 
                           ~ggplot(.x, aes(UMAP1,UMAP2, color = Condition)) + 
                             geom_point(size = 3.5) +
                             theme_bw(base_size = 14) + 
                             labs(title = .y) +
                             scale_color_manual(values = c("black","grey70")) +
                             theme(panel.grid = element_blank()))
#walk2(umap_with_elements, names(umap_with_elements),
 #     ~ggsave(filename = here::here(str_c("final results/UMAP with axis elements ",.y,".png")), plot = .x, dpi = 1080, width = 6, height = 4))
umap_blank <- map2(umap_df_list, names(umap_df_list), 
                   ~ggplot(.x, aes(UMAP1,UMAP2, color = Condition)) + 
                     geom_point(size = 3.5) +
                     labs(title = .y, x = " ", y = " ") +
                     theme_void(base_size = 16) + 
                     scale_color_manual(values = c("black","grey70")) +
                     theme(panel.border = element_rect(linewidth = 1, color = "grey50", fill = NA),
                           panel.grid = element_blank(),
                           legend.title = element_blank(),
                           axis.title = element_text(),
                           plot.background = element_rect(fill = "white")))
#walk2(umap_blank, names(umap_blank),
 #     ~ggsave(filename = here::here(str_c("final results/UMAP blank ",.y,".png")), plot = .x, dpi = 1080, width = 6, height = 4))
```


# Session Info
```{r eval=FALSE}
sessionInfo()
```


