---
title: "Visualizing Bile Acids Proteomics results"
output: html_notebook
---

```{r lib-setup, echo=FALSE, include=FALSE}
library(clusterProfiler)
library(EnhancedVolcano)
library(org.Hs.eg.db)
library(tidyverse)
library(readxl)
```

```{r import}
mq <- read_tsv(here::here("normalyzer_limma_de_results_annotated.tsv"))
meta <- read_tsv(here::here("norm_meta.tsv"))
mq_res <- mq %>% 
  select(contains("Log"), contains("Adj"), contains("Majority"), contains("Gene"))
mq_mat <- mq %>% 
  select(`Gene names`, `Majority protein IDs`, matches("(C|L)[1-6](c|m)"))
#Convert the results table into a list and unify the columns names. Invert the foldChanges, so that comparisons are L-C instead of C-L.
#Reduce Gene names to just one gene for analysis
mq_res_list <- list(select(mq_res, -contains("Cc-Lc")),
                    select(mq_res, -contains("Cm-Lm"))) %>% 
  set_names(nm = c("Membrane", "Cyto")) %>% 
  map(rename_with, ~str_remove(.,"C(c|m)-L(c|m)_") ,matches("C(c|m)-L(c|m)_")) %>% 
  map(mutate, across(contains("log2"), ~ -.x)) %>% 
  map(separate, `Gene names`, into = letters [1:2], extra = "drop", sep = ";") %>% 
  map(mutate, Genes = case_when(
    str_detect(a, "orf") & !is.na(b) & str_detect(b, "orf", negate = T) ~ b,
    TRUE ~ a
  ), Gene_rows = Genes) 
```

Need to make the following:

    A Venn diagram depicting number for differentially abundant proteins identified in cytoplasmic fraction, in membrane fraction or in both the fraction
    Volcano plots
    And the dot plots from GO ontogeny analysis just like the link you sent.

# Venn diagram

```{r venn}
#Filter results to only include DA proteins
da_list <- mq_res_list %>% 
  map(filter, AdjPVal < 0.05, abs(log2FoldChange) > 1)
#Pull just the Protein IDs for DA proteins
da_genes <- da_list %>% map(pull, `Majority protein IDs`) %>% 
  set_names(nm = c("Membrane-bound proteins", "Cytosolic proteins"))
ggvenn::ggvenn(da_genes, text_size = 10, set_name_size = 6,
               fill_color = c("indianred", "royalblue1"), show_percentage = F) 
#ggsave(here::here("Venn diagram DA proteins.png"), units = "in", dpi = 720, width = 8, height = 6)
```


# Volcano plots
```{r enhancedvolcano, fig.width=10}
df <- mq_res_list[[1]]
rows <- map(mq_res_list, pull, var = Gene_rows)
titles <- names(da_genes)
top_hits <- mq_res_list %>% 
  map(filter, AdjPVal < 0.05) %>% 
  map(arrange, desc(abs(log2FoldChange))) %>% 
  map(slice, 1:7) %>% 
  map(pull, var = Gene_rows)
plot_list <- list(df = mq_res_list, rows = rows, title = titles, selectLab = top_hits)
#Create a custom volcano plot function
customvolcano <- function(df, rows, title, selectLab="") {
  print(EnhancedVolcano(toptable = df,lab = rows, x = "log2FoldChange", y = "AdjPVal",
                pCutoff = 0.05, FCcutoff = 1, legendPosition = 'bottom', title = title, subtitle = NULL,
                selectLab = selectLab, drawConnectors = T, boxedLabels = T))
}
#pdf(here::here("Volcano plots top 7 hits labelled.pdf"), width = 10)
pwalk(plot_list, customvolcano)
#dev.off()
```
# Functional enrichment
```{r cluster-enrich}
set.seed(711)
#Get the whole human genome as universe
human_genom <- org.Hs.egSYMBOL
mapped_genes <- mappedRkeys(human_genom)
h <- read.gmt(list.files(path = here::here("."),pattern = "h.all"))
c2_cp <- read.gmt(list.files(path = here::here("."),pattern = "c2.cp"))  
c2_noID <- c2_cp %>% 
  mutate(term = str_remove(term, "[:alpha:]+_"))
#Set up lists of genes for analysis
#Pull just the gene names for DA proteins
da_symbols <- da_list %>% map(pull, `Genes`) %>% 
  set_names(nm = c("Membrane-bound proteins", "Cytosolic proteins"))
enrich_h_list <- map(da_symbols, ~enricher(.,
                     TERM2GENE = h,
                     universe = mapped_genes,
                     pAdjustMethod = "BH",
                     pvalueCutoff = 0.05,
                     qvalueCutoff = 0.05))
enrich_c2_list <- map(da_symbols, ~enricher(.,
                     TERM2GENE = c2_noID,
                     universe = mapped_genes,
                     pAdjustMethod = "BH",
                     pvalueCutoff = 0.05,
                     qvalueCutoff = 0.05))
ego_bp <- map(da_symbols, ~enrichGO(gene = .,
                OrgDb         = org.Hs.eg.db,
                universe = mapped_genes,
                keyType       = 'SYMBOL',
                ont           = "BP",
                pAdjustMethod = "BH",
                pvalueCutoff  = 0.05,
                qvalueCutoff  = 0.05))
ego_bp <- map(ego_bp,enrichplot::pairwise_termsim)
ego_bp2 <- ego_bp %>% 
  map(~clusterProfiler::simplify(., cutoff=0.4, by="p.adjust", select_fun=min))
```

Plot the functional enrichment results

```{r cluster-plotting, fig.height=9, fig.width=7}
dot_titles <- rep(c("Hallmark geneset", "C2 Canonical Pathways", "GO BP"), each = 2) %>% 
  str_c(names(da_genes), sep = "\n")
#pdf(here::here("Functional enrichment analyses.pdf"), height = 9, width = 7)
walk2(c(enrich_h_list, enrich_c2_list, ego_bp2), dot_titles, 
      ~print(dotplot(.x, showCategory = 15) + ggtitle(.y)))
#dev.off()
```


